File: test/core/privacy/version_parser_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:luvi_app/core/privacy/version_parser.dart';

void main() {
  group('VersionParser.parseMajorVersion', () {
    test('parses v{major} format', () {
      expect(VersionParser.parseMajorVersion('v1'), equals(1));
      expect(VersionParser.parseMajorVersion('v2'), equals(2));
      expect(VersionParser.parseMajorVersion('v10'), equals(10));
    });

    test('parses v{major}.{minor} format', () {
      expect(VersionParser.parseMajorVersion('v1.0'), equals(1));
      expect(VersionParser.parseMajorVersion('v2.5'), equals(2));
      expect(VersionParser.parseMajorVersion('v10.99'), equals(10));
    });

    test('throws on invalid format', () {
      expect(
        () => VersionParser.parseMajorVersion('1.0'),
        throwsFormatException,
      );
      expect(
        () => VersionParser.parseMajorVersion('version1'),
        throwsFormatException,
      );
      expect(
        () => VersionParser.parseMajorVersion('v1.0.1'),
        throwsFormatException,
      );
      expect(
        () => VersionParser.parseMajorVersion(''),
        throwsFormatException,
      );
    });
  });

  group('VersionParser.parseMinorVersion', () {
    test('returns 0 for v{major} format', () {
      expect(VersionParser.parseMinorVersion('v1'), equals(0));
      expect(VersionParser.parseMinorVersion('v2'), equals(0));
    });

    test('parses minor from v{major}.{minor}', () {
      expect(VersionParser.parseMinorVersion('v1.0'), equals(0));
      expect(VersionParser.parseMinorVersion('v1.5'), equals(5));
      expect(VersionParser.parseMinorVersion('v2.99'), equals(99));
    });

    test('throws on invalid format', () {
      expect(
        () => VersionParser.parseMinorVersion('bad'),
        throwsFormatException,
      );
    });
  });

  group('VersionParser.isValidFormat', () {
    test('returns true for valid formats', () {
      expect(VersionParser.isValidFormat('v1'), isTrue);
      expect(VersionParser.isValidFormat('v1.0'), isTrue);
      expect(VersionParser.isValidFormat('v10.99'), isTrue);
    });

    test('returns false for invalid formats', () {
      expect(VersionParser.isValidFormat('1.0'), isFalse);
      expect(VersionParser.isValidFormat('version1'), isFalse);
      expect(VersionParser.isValidFormat('v1.0.1'), isFalse);
      expect(VersionParser.isValidFormat(''), isFalse);
      expect(VersionParser.isValidFormat('v'), isFalse);
    });
  });
}

File: supabase/functions/_shared/version_parser.test.ts
import { assertEquals } from "https://deno.land/std@0.224.0/assert/mod.ts";
import { parseVersion, isValidVersionFormat } from "./version_parser.ts";

Deno.test("parseVersion: valid v{major} format", () => {
  const result1 = parseVersion("v1");
  assertEquals(result1.valid, true);
  assertEquals(result1.major, 1);
  assertEquals(result1.minor, 0);

  const result2 = parseVersion("v10");
  assertEquals(result2.valid, true);
  assertEquals(result2.major, 10);
  assertEquals(result2.minor, 0);
});

Deno.test("parseVersion: valid v{major}.{minor} format", () => {
  const result1 = parseVersion("v1.0");
  assertEquals(result1.valid, true);
  assertEquals(result1.major, 1);
  assertEquals(result1.minor, 0);

  const result2 = parseVersion("v2.5");
  assertEquals(result2.valid, true);
  assertEquals(result2.major, 2);
  assertEquals(result2.minor, 5);
});

Deno.test("parseVersion: invalid formats", () => {
  const invalid = ["1.0", "version1", "v1.0.1", "", "v", "bad"];
  for (const input of invalid) {
    const result = parseVersion(input);
    assertEquals(result.valid, false);
    assertEquals(result.error !== undefined, true);
  }
});

Deno.test("isValidVersionFormat: validates correctly", () => {
  assertEquals(isValidVersionFormat("v1"), true);
  assertEquals(isValidVersionFormat("v1.0"), true);
  assertEquals(isValidVersionFormat("v10.99"), true);

  assertEquals(isValidVersionFormat("1.0"), false);
  assertEquals(isValidVersionFormat("version1"), false);
  assertEquals(isValidVersionFormat("v1.0.1"), false);
  assertEquals(isValidVersionFormat(""), false);
});
